core/src/net/cidr/v4.rs

```diff
@@ -77,33 +170,40 @@ impl Cidr for Ipv4Cidr {
     }
 
     #[inline]
-    fn contains(&self, address: Self::Addr) -> bool {
-        self.prefix == (self.mask & u32::from_be_bytes(address.octets()))
+    fn size(&self) -> usize {
+        2usize.pow((IPV4ADDR_BITS - self.length) as u32)
     }
 
     #[inline]
-    fn contains_ip(&self, ip: IpAddr) -> bool {
-        match ip {
-            IpAddr::V4(ip) => self.contains(ip),
-            _ => false,
-        }
+    fn with_netmask(address: Self::Addr, netmask: Self::Addr) -> Result<Self, CidrError> {
+        let length = Ipv4Cidr::netmask_length(netmask)?;
+        Ipv4Cidr::new(address, length)
     }
 }
 
 impl FromStr for Ipv4Cidr {
-    type Err = CidrParseError;
+    type Err = CidrError;
 
-    fn from_str(s: &str) -> Result<Self, CidrParseError> {
+    fn from_str(s: &str) -> Result<Self, CidrError> {
         match s.split('/').collect::<Vec<&str>>().as_slice() {
-            [addr, len] => {
+            [addr, len_or_netmask] => {
                 let address =
-                    Ipv4Addr::from_str(addr).map_err(|e| CidrParseError(e.to_string()))?;
-                let length =
-                    usize::from_str_radix(len, 10).map_err(|e| CidrParseError(e.to_string()))?;
+                    Ipv4Addr::from_str(addr).map_err(|e| CidrError::ParseFailure(e.to_string()))?;
 
-                Ipv4Cidr::new(address, length)
+                if let Ok(len) = usize::from_str_radix(len_or_netmask, 10) {
+                    Ipv4Cidr::new(address, len)
+                } else {
+                    let netmask = Ipv4Addr::from_str(len_or_netmask).map_err(|e| {
+                        CidrError::ParseFailure(format!(
```

Comment:
should let the error handle the formatting. this will generate very weird looking error text because it's being formatted here then reformatted again by the attribute.